File: pipelines/imperative_pandas/categorize_endpoint_type.py
from __future__ import annotations

import pandas as pd


IMAGE_EXTENSIONS = (".gif", ".jpg", ".jpeg", ".png", ".bmp", ".svg", ".ico", ".webp", ".tif", ".tiff")
DOWNLOAD_EXTENSIONS = (".zip", ".gz", ".tar", ".tgz", ".bz2", ".7z", ".rar", ".exe", ".dmg", ".pdf")
HTML_EXTENSIONS = (".html", ".htm")


def classify(endpoint: str) -> str:
    e = (endpoint or "").lower()
    if e == "/" or e.endswith("/") or e.endswith(HTML_EXTENSIONS):
        return "html"
    if e.endswith(IMAGE_EXTENSIONS):
        return "image"
    if e.endswith(DOWNLOAD_EXTENSIONS):
        return "download"
    if "/cgi-bin/" in e or e.endswith(".cgi") or "?" in e or "=" in e:
        return "cgi"
    return "other"


def run(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()
    out["endpoint_type"] = out["endpoint"].astype(str).map(classify)
    return out

File: pipelines/functional_ocaml/categorize_endpoint_type.ml
open Etl

let ends_with value suffix =
  let lv = String.length value in
  let ls = String.length suffix in
  lv >= ls && String.equal (String.sub value (lv - ls) ls) suffix

let lowercase = String.lowercase_ascii

let is_html endpoint =
  endpoint = "/"
  || List.exists (fun ext -> ends_with endpoint ext) [ ".html"; ".htm"; "/" ]

let is_image endpoint =
  List.exists
    (fun ext -> ends_with endpoint ext)
    [ ".gif"; ".jpg"; ".jpeg"; ".png"; ".bmp"; ".svg"; ".ico"; ".webp"; ".tif"; ".tiff" ]

let is_download endpoint =
  List.exists
    (fun ext -> ends_with endpoint ext)
    [ ".zip"; ".gz"; ".tar"; ".tgz"; ".bz2"; ".7z"; ".rar"; ".exe"; ".dmg"; ".pdf" ]

let is_cgi endpoint = String.contains endpoint '?' || String.contains endpoint '=' || ends_with endpoint ".cgi"

let classify endpoint =
  let endpoint = lowercase endpoint in
  if is_html endpoint then "html"
  else if is_image endpoint then "image"
  else if is_download endpoint then "download"
  else if is_cgi endpoint then "cgi"
  else "other"

let apply row = Row.set "endpoint_type" (classify (Row.get_exn "endpoint" row)) row
